program/shell/errors_to_json.go

```diff
@@ -0,0 +1,46 @@
+package shell
+
+import (
+	"encoding/json"
+
+	"github.com/elves/elvish/eval"
+	"github.com/elves/elvish/parse"
+)
+
+// An auxiliary struct for converting errors with diagnostics information to JSON.
+type errorInJSON struct {
+	FileName string `json:"fileName"`
+	Start    int    `json:"start"`
+	End      int    `json:"end"`
+	Message  string `json:"message"`
+}
+
+// An auxiliary struct for converting errors with only a message to JSON.
+type simpleErrorInJSON struct {
+	Message string `json:"message"`
+}
+
+// ErrorToJSON converts the error into JSON format.
+func ErrorToJSON(err error) []byte {
+	var e interface{}
+	switch err := err.(type) {
+	case *eval.CompilationError:
+		e = []interface{}{
+			errorInJSON{err.Context.Name, err.Context.Begin, err.Context.End, err.Message},
+		}
+	case *parse.MultiError:
+		var errArr []errorInJSON
+		for _, v := range err.Entries {
+			errArr = append(errArr,
+				errorInJSON{v.Context.Name, v.Context.Begin, v.Context.End, v.Message})
+		}
+		e = errArr
+	default:
+		e = []interface{}{simpleErrorInJSON{err.Error()}}
+	}
+	jsonError, er := json.Marshal(e)
```

Comment:
Variables that keep errors are always called `err` or some name starting with `err`. Since `err` is taken in the function, call the second variable `errMarshal`.